Userland/Libraries/LibGfx/TextLayout.cpp

```diff
@@ -12,7 +12,7 @@
 
 namespace Gfx {
 
-RefPtr<GlyphRun> shape_text(FloatPoint baseline_start, Utf8View string, Gfx::Font const& font, GlyphRun::TextType text_type)
+RefPtr<GlyphRun> shape_text(FloatPoint baseline_start, float x_offset, float tab_width, Utf8View string, Gfx::Font const& font, GlyphRun::TextType text_type)
```

Comment:
what is x_offset and why a function responsible for text shaping should be aware of it? there is already a baseline_start that is used to encode shaped text position.

Reply:
It's a fair point. I was using it to measure how far the first glyph is from the left edge of the containing block to measure where the next tab stop will be. 

https://drafts.csswg.org/css-text/#tab-size-property

I'll only need to use this if there is indeed a tab in the text chunk. 

http://wpt.live/css/css-text/white-space/tab-stop-threshold-002.html

An example is this test where for instance a span can measure 7.6ch, pushing the first glyph out, the tab-size is 8 the distance is the (8 - 7.6) = 0.4ch however according to the spec we have to move this to the _next_ tab stop as the distance is <0.5ch. 

So given these types of calculations this was my first attempt on how to pass this data in the shape text run so we could shift the glyph.

Reply:
> measure how far the first glyph is from the left edge of the containing block

hmm, offset of the left edge of containing block is always zero, no?

I checked how x_offset is currently calculated and this code won't get you left edge:
```cpp
auto boxes = m_containing_block_used_values.line_boxes;
CSSPixels accumulated_width = 0;
for (auto& box : boxes) {
    accumulated_width += box.width();
}
```
if containing block has two lines of widths 100px and 50px this code will give you 150px and that's definitely not what you want here.

you can use this html as a test. it's currently not rendered correctly with your changes:
```html
<!DOCTYPE html>
<style>
.tab-size {
    white-space: pre;
    background-color: #f0f0f0;
    padding: 10px;
    tab-size: 20;
}
</style>
<div class="tab-size">
    <code>
function hello() {<br>
	let	greeting = "Hello, world!";<br>
	console.log(greeting);<br>
}
    </code>
</div>
```

Reply:
By offset I mean where the text fragment get's rendered relative to the containing block. So in your example, I would want to know the 150px as I would have to align the next glyph after encountering a tab to the closest tab taking that width into consideration. My variable is probably confusingly misnamed :)

not sure about your example, but I think this [wpt test](https://wpt.live/css/css-text/white-space/tab-stop-threshold-002.html) demonstrates it well, see the code below for the source and explanation (as I understand it):
```html
<style>
div {
  white-space: pre;
  font-family: monospace;
  tab-size: 8; /* the initial value, but since we're measuring against it, we might as well be sure */
}
span { display: inline-block; }
#s1 { width: 7.6ch; }
#s2 { width: 7.8ch; }
#s3 { width: 7.9ch; }
</style>

<p>Test passes if the 4 letters bellow are vertically aligned.
<div><span id=s1></span>&#9;P</div>
<div><span id=s2></span>&#9;A</div>
<div><span id=s3></span>&#9;S</div>
<div>                S</div>
```

Here start counting tab stops at the tab-size, which is multiples of 8 from the left edge of the div element (I believe this is the containing block of the text fragment). The span element pushes the text fragment to start at 7.6ch from the left edge of the div. So closest tab stop is at 8, at a distance of (8 - 7.6) = 0.4. 

However according to the spec, if the distance is <0.5ch we have to push it to the _next_ _next_ tab stop, in this instance at position 16, so a translation of 8.4 of the next glyph after the tab which also takes into account the span width.

So during text shaping when we encounter a tab we would somehow need to know if the text has been shifted by any elements before it and account for that when we translate the glyph to the next tab stop.

In the example, the last div text should align to the tab shifted text which would be equivalent to 16 spaces, hence testing the <0.5ch in which we have to move the glyph after the tab not to position 8 but rather 16.

Hope this makes sense. Like I said unsure if there is a way to retrieve this information without having to add parameters to the shaping function.

Reply:
> By offset I mean where the text fragment get's rendered relative to the containing block. So in your example, I would want to know the 150px as I would have to align the next glyph after encountering a tab to the closest tab taking that width into consideration. My variable is probably confusingly misnamed :)

I might actually be missing something but using a sum of lines width still does not make sense.

for example:
```
> first line: abc (width = 100px)
> second line: a (width = 50px)
```

why would you want to know sum of all lines widths during text shaping if each line has its own (or multiple) glyph runs? I can imagine how max() of line widths could be applied, but why sum?

Reply:
I guess put another way how would you recommend finding the position to shift the glyph per the spec

> If the tab size is zero, preserved tabs are not rendered. Otherwise, each preserved tab is rendered as a horizontal shift that lines up the start edge of the next glyph with the next tab stop. If this distance is less than 0.5ch, then the subsequent tab stop is used instead. Tab stops occur at points that are multiples of the tab size from the starting content edge of the preserved tab’s nearest block container ancestor. The tab size is given by the tab-size property

How can I figure out the distance to the next tab stop given where the glyph sits before the encountered tab relative to its block container?

I thought summing up the line boxes (and I guess any widths of the glyphs encountered before the tab) would tell me the accumulated width up until the glyph and then I just have to shift it the difference of the tab multiple and that width to align it to the next tab stop. I could be misunderstanding something fundamental about layout here though!